src/datanode/src/heartbeat/handler/upgrade_region.rs

```diff
@@ -22,115 +25,148 @@ use crate::heartbeat::task_tracker::WaitResult;
+                        location_id,
+                        checkpoint,
+                    },
+                ));
             }
-            WaitResult::Finish(Err(err)) => {
-                Some(InstructionReply::UpgradeRegion(UpgradeRegionReply {
-                    ready: false,
-                    exists: true,
-                    error: Some(format!("{err:?}")),
-                }))
+        }
+
+        let mut wait_results = Vec::with_capacity(catchup_request.len());
+
+        for (region_id, replay_timeout, catchup_request) in catchup_request {
+            let region_server_moved = ctx.region_server.clone();
+            // TODO(weny): parallelize the catchup tasks.
+            let result = ctx
+                .catchup_tasks
+                .try_register(
+                    region_id,
+                    Box::pin(async move {
+                        info!(
+                            "Executing region: {region_id} catchup to: last entry id {:?}",
+                            catchup_request.entry_id
+                        );
+                        region_server_moved
+                            .handle_request(region_id, RegionRequest::Catchup(catchup_request))
+                            .await?;
+                        Ok(())
+                    }),
+                )
+                .await;
+
+            if result.is_busy() {
+                warn!("Another catchup task is running for the region: {region_id}");
             }
+
+            // We don't care that it returns a newly registered or running task.
+            let mut watcher = result.into_watcher();
+            wait_results.push((
+                region_id,
+                ctx.catchup_tasks.wait(&mut watcher, replay_timeout).await,
+            ));
         }
+
+        let results = join_all(
+            wait_results
+                .into_iter()
+                .map(|(region_id, result)| async move {
+                    match result {
+                        WaitResult::Timeout => UpgradeRegionReply {
+                            region_id,
+                            ready: false,
+                            exists: true,
+                            error: None,
+                        },
+                        WaitResult::Finish(Ok(_)) => UpgradeRegionReply {
+                            region_id,
+                            ready: true,
+                            exists: true,
+                            error: None,
+                        },
+                        WaitResult::Finish(Err(err)) => UpgradeRegionReply {
+                            region_id,
+                            ready: false,
+                            exists: true,
+                            error: Some(format!("{err:?}")),
+                        },
+                    }
+                }),
+        )
+        .await;
+
+        replies.extend(results.into_iter());
+        replies
```

Comment:
So we don't guarantee the replies has the same order as `upgrade_regions`?

Reply:
No, each reply includes a region_id field, so the order is not guaranteed.